src/shims/env.rs

```diff
@@ -29,42 +30,100 @@ impl<'tcx> EnvVars<'tcx> {
             for (name, value) in env::vars() {
                 if !excluded_env_vars.contains(&name) {
                     let var_ptr =
-                        alloc_env_var_as_c_str(name.as_ref(), value.as_ref(), ecx);
+                        alloc_env_var_as_target_str(name.as_ref(), value.as_ref(), ecx)?;
                     ecx.machine.env_vars.map.insert(OsString::from(name), var_ptr);
                 }
             }
         }
         ecx.update_environ()
     }
+
+    pub(super) fn values(&self) -> InterpResult<'tcx, Values<'_, OsString, Pointer<Tag>>> {
+        Ok(self.map.values())
+    }
 }
 
-fn alloc_env_var_as_c_str<'mir, 'tcx>(
+fn alloc_env_var_as_target_str<'mir, 'tcx>(
     name: &OsStr,
     value: &OsStr,
     ecx: &mut InterpCx<'mir, 'tcx, Evaluator<'tcx>>,
-) -> Pointer<Tag> {
+) -> InterpResult<'tcx, Pointer<Tag>> {
     let mut name_osstring = name.to_os_string();
     name_osstring.push("=");
     name_osstring.push(value);
-    ecx.alloc_os_str_as_c_str(name_osstring.as_os_str(), MiriMemoryKind::Machine.into())
+    Ok(ecx.alloc_os_str_as_target_str(name_osstring.as_os_str(), MiriMemoryKind::Machine.into())?)
 }
 
 impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
 pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
-    fn getenv(&mut self, name_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
-        let this = self.eval_context_mut();
+    fn getenv(&self, name_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_ref();
 
         let name_ptr = this.read_scalar(name_op)?.not_undef()?;
-        let name = this.read_os_str_from_c_str(name_ptr)?;
-        Ok(match this.machine.env_vars.map.get(name) {
-            // The offset is used to strip the "{name}=" part of the string.
+        let name = this.read_os_str_from_target_str(name_ptr)?;
+        Ok(match this.machine.env_vars.map.get(&name) {
             Some(var_ptr) => {
+                // The offset is used to strip the "{name}=" part of the string.
                 Scalar::from(var_ptr.offset(Size::from_bytes(u64::try_from(name.len()).unwrap().checked_add(1).unwrap()), this)?)
             }
             None => Scalar::ptr_null(&*this.tcx),
         })
     }
 
+
+    fn getenvironmentvariablew(
+        &mut self,
+        name_op: OpTy<'tcx, Tag>, // LPCWSTR lpName
+        buf_op: OpTy<'tcx, Tag>, // LPWSTR  lpBuffer
+        size_op: OpTy<'tcx, Tag>, // DWORD   nSize
+    ) -> InterpResult<'tcx, u32> {
+        let this = self.eval_context_mut();
+
+        let name_ptr = this.read_scalar(name_op)?.not_undef()?;
+        let name = this.read_os_str_from_target_str(name_ptr)?;
```

Comment:
Since this is a Windows-specific function, please (a) call `this.assert_target_os("windows", "GetEnvironmentVariableW")`, and (b) use `read_os_str_from_wide_str` here. The rest of the code anyway would not work for other targets.